sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -0,0 +1,1606 @@
+ * could not fit in the SpillableHostStore's limit (if defined). In this case, the call to spill
+ * will either create a host handle tracking an object on the host store (if we made room), or it
+ * will create a host handle that points to a disk handle, tracking a file on disk.
+ *
+ * If the disk is full, we do not handle this in any special way. We expect this to be a
+ * terminal state to the executor. Every handle spills to its own file on disk, identified
+ * as a "temporary block" `BlockId` from Spark.
+ *
+ * Notes on locking:
+ *
+ * All stores use a concurrent hash map to store instances of `StoreHandle`. The only store
+ * with extra locking is the `SpillableHostStore`, to maintain a `totalSize` number that is
+ * used to figure out cheaply when it is full.
+ *
+ * Handles hold a lock to protect the user against when it is either in the middle of
+ * spilling, or closed. Most handles, except for disk handles, hold a reference to an object
+ * in their respective store (`SpillableDeviceBufferHandle` has a `dev` reference that holds
+ * a `DeviceMemoryBuffer`, and a `host` reference to `SpillableHostBufferHandle` that is only
+ * set if spilled), and the handle guarantees that one of these will be set, or none if the handle
+ * is closed.
+ *
+ * We hold the handle lock when we are spilling (performing IO). That means that no other consumer
+ * can access this spillable device handle while it is being spilled, including a second thread
+ * trying to spill and asking each of the handles whether they are spillable or not, as that
+ * requires the handle lock. We will relax this likely in follow on work.
+ *
+ * We never hold a store-wide coarse grain lock in the stores when we do IO.
+ */
+
+/**
+ * Common interface for all handles in the spill framework.
+ */
+trait StoreHandle extends AutoCloseable {
+  /**
+   * Approximate size of this handle, used in two scenarios:
+   * - Used from the host store to figure out how much host memory total it is tracking.
+   * - If sizeInBytes is 0, the object is tracked by the stores so it can be
+   *   removed on shutdown, or by handle.close, but 0-byte handles are not spillable.
+   */
+  val sizeInBytes: Long
+}
+
+trait SpillableHandle extends StoreHandle {
+  /**
+   * Method called to spill this handle. It can be triggered from the spill store,
+   * or directly against the handle.
+   * @return sizeInBytes if spilled, 0 for any other reason (not spillable, closed)
+   */
+  def spill: Long
+
+  /**
+   * Method used to determine whether a handle tracks an object that could be spilled
+   * @note At the level of `SpillableHandle`, the only requirement of spillability
+   *       is that the size of the handle is > 0. `sizeInBytes` is known at construction, and
+   *       is a `val`.
+   * @return true if currently spillable, false otherwise
+   */
+  private[spill] def spillable: Boolean = sizeInBytes > 0
```

Comment:
Why is this private to spill? Just curious, because if we want others to add spillable things in does that mean they have to put the handles in com.nvidia.spark.rapids.spill to make them work?

Reply:
I think so. We can try to open that up if we want spillables in other packages, but right now it's all in here so I made this change in response to a comment by @jlowe. I did want to access this from unit test, and that's why it is `private[spill]` specifically.